src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -303,34 +303,24 @@ bool TryGetFileMappingFromMetadataImporter(FileKey fileKey, [NotNullWhen(true)]
         FileKey fileKey,
         Func<FileKey, (ModuleMetadata moduleMetadata, TemporaryStorageStreamHandle? storageHandle)> moduleMetadataFactory)
     {
-        var (manifestModule, manifestHandle) = moduleMetadataFactory(fileKey);
-
         using var _1 = ArrayBuilder<ModuleMetadata>.GetInstance(out var moduleBuilder);
         using var _2 = ArrayBuilder<TemporaryStorageStreamHandle?>.GetInstance(out var storageHandles);
 
-        string? assemblyDir = null;
+        var (manifestModule, manifestHandle) = moduleMetadataFactory(fileKey);
+        moduleBuilder.Add(manifestModule);
+        storageHandles.Add(manifestHandle);
+
+        var assemblyDir = Path.GetDirectoryName(fileKey.FullPath);
         foreach (var moduleName in manifestModule.GetModuleNames())
         {
-            if (assemblyDir is null)
-            {
-                moduleBuilder.Add(manifestModule);
-                assemblyDir = Path.GetDirectoryName(fileKey.FullPath);
-            }
-
             // Suppression should be removed or addressed https://github.com/dotnet/roslyn/issues/41636
             var moduleFileKey = FileKey.Create(PathUtilities.CombineAbsoluteAndRelativePaths(assemblyDir, moduleName)!);
-            var (metadata, metadataStorageHandle) = moduleMetadataFactory(moduleFileKey);
 
+            var (metadata, metadataStorageHandle) = moduleMetadataFactory(moduleFileKey);
             moduleBuilder.Add(metadata);
             storageHandles.Add(metadataStorageHandle);
         }
 
-        if (moduleBuilder.Count == 0)
```

Comment:
Second, if there were *no* submodules, we'd add the primary module here *and* the handle.  But if there *were* submodules, then the handle for the primary module was dropped.